File: docs/recipes/generators/python/json_schema.py
"""EXAMPLE GENERATOR — copy into your project as codegen/generators/json_schema.py and own it.

A worked example, NOT a supported MetaObjects product surface: no release promises its output.

Wire it as ``module:symbol`` — in metaobjects.config.yaml:

    targets:
      api:
        outDir: gen
        generators: [codegen.generators.json_schema:json_schema, codegen.generators.json_schema:openapi]

or on the command line: ``metaobjects gen ./metaobjects --out gen --generators
codegen.generators.json_schema:json_schema``. ``metaobjects verify --codegen`` re-runs it with
the same selection and fails when committed output is stale. (codegen/ and
codegen/generators/ need an ``__init__.py`` each so the module imports.)

Emits schemas/<package path>/<Name>.schema.json per concrete object (JSON Schema 2020-12) and
openapi.json (OpenAPI 3.1) with the cross-port CRUD paths for each object that has a source.
"""
from __future__ import annotations

import json
from typing import Any, Callable

from metaobjects.codegen.model_walk import (
    EmittedFile,
    GenContext,
    description,
    enum_values,
    field_is_array,
    is_abstract,
    is_required,
    max_length,
    object_ref_target,
    package_of,
    per_entity,
    per_model,
    primary_key_fields,
    route_path,
)

# field subtype -> JSON Schema, following the cross-port wire contract
# (docs/features/api-contract.md, "Type encodings").
_SCALARS: dict[str, dict[str, Any]] = {
    "string": {"type": "string"},
    "int": {"type": "integer"},
    "long": {"type": "integer"},
    "currency": {"type": "integer"},
    "double": {"type": "number"},
    "float": {"type": "number"},
    "decimal": {"type": "string", "pattern": "^-?\\d+(\\.\\d+)?$"},
    "boolean": {"type": "boolean"},
    "date": {"type": "string", "format": "date"},
    "time": {"type": "string", "format": "time"},
    "timestamp": {"type": "string", "format": "date-time"},
    "uuid": {"type": "string", "format": "uuid"},
    "uri": {"type": "string", "format": "uri"},
    "map": {"type": "object"},
}


def _dump(doc: dict[str, Any]) -> str:
    return json.dumps(doc, indent=2) + "\n"


def schema_path(obj: Any) -> str:
    pkg = package_of(obj)
    return f"schemas/{pkg.replace('::', '/')}/{obj.name}.schema.json" if pkg else f"schemas/{obj.name}.schema.json"


def field_schema(field: Any, ref_for: Callable[[Any], str]) -> dict[str, Any]:
    if field.sub_type == "enum":
        value: dict[str, Any] = {"type": "string", "enum": enum_values(field)}
    elif field.sub_type == "object":
        target = object_ref_target(field)
        value = {"$ref": ref_for(target)} if target is not None else {"type": "object"}
    else:
        value = dict(_SCALARS.get(field.sub_type, {"type": "string"}))
    if (n := max_length(field)) is not None:
        value["maxLength"] = n
    # field_is_array RESOLVES; the raw field.is_array flag misses an inherited isArray.
    schema = {"type": "array", "items": value} if field_is_array(field) else value
    if (d := description(field)) is not None:
        schema["description"] = d
    return schema


def object_schema(obj: Any, ref_for: Callable[[Any], str]) -> dict[str, Any]:
    # fields() RESOLVES: an object that `extends` a base gets the base's fields too.
    fields = obj.fields()
    schema: dict[str, Any] = {
        "title": obj.name,
        "type": "object",
        "properties": {f.name: field_schema(f, ref_for) for f in fields},
    }
    required = [f.name for f in fields if is_required(f)]
    if required:
        schema["required"] = required
    if (d := description(obj)) is not None:
        schema["description"] = d
    return schema


class JsonSchemaGenerator:
    """One JSON Schema per concrete object. A generator is just ``name`` + ``generate``."""

    name = "json-schema"

    def filter(self, obj: Any) -> bool:
        # ctx.entities is EVERY object, abstract bases included; the runner applies this
        # as ctx.matches, which per_entity honours.
        return not is_abstract(obj)

    def generate(self, ctx: GenContext) -> list[EmittedFile]:
        def one(obj: Any, _ctx: GenContext) -> EmittedFile:
            def ref(target: Any) -> str:  # same directory for one package; adjust if yours differ
                return f"./{target.name}.schema.json"

            doc = {"$schema": "https://json-schema.org/draft/2020-12/schema", "$id": schema_path(obj),
                   **object_schema(obj, ref)}
            return EmittedFile(path=schema_path(obj), content=_dump(doc))

        return per_entity(one)(ctx)


class OpenApiGenerator:
    """One OpenAPI 3.1 document for the whole model."""

    name = "openapi"

    def __init__(self, title: str = "API", api_prefix: str = "") -> None:
        self.title = title
        self.api_prefix = api_prefix

    def filter(self, obj: Any) -> bool:
        return not is_abstract(obj)

    def generate(self, ctx: GenContext) -> list[EmittedFile]:
        def whole(objects: list[Any], _ctx: GenContext) -> EmittedFile:
            def ref(target: Any) -> str:
                return f"#/components/schemas/{target.name}"

            schemas = {o.name: object_schema(o, ref) for o in objects}
            paths: dict[str, Any] = {}
            for o in objects:
                # children() RESOLVES, so an inherited source counts; a value object has none.
                if not any(c.type == "source" for c in o.children()):
                    continue
                base = f"{self.api_prefix}/{route_path(o.name)}"
                row = {"$ref": ref(o)}
                body = {"content": {"application/json": {"schema": row}}}
                pk = primary_key_fields(o)
                pk_field = next((f for f in o.fields() if pk and f.name == pk[0]), None)
                id_param = {"name": "id", "in": "path", "required": True,
                            "schema": field_schema(pk_field, ref) if pk_field else {"type": "string"}}
                paths[base] = {
                    "get": {"operationId": f"list{o.name}", "responses": {"200": {
                        "description": "OK",
                        "content": {"application/json": {"schema": {"type": "array", "items": row}}}}}},
                    "post": {"operationId": f"create{o.name}", "requestBody": body,
                             "responses": {"201": {"description": "Created", **body}}},
                }
                paths[f"{base}/{{id}}"] = {
                    "parameters": [id_param],
                    "get": {"operationId": f"get{o.name}", "responses": {"200": {"description": "OK", **body}}},
                    "patch": {"operationId": f"update{o.name}", "requestBody": body,
                              "responses": {"200": {"description": "OK", **body}}},
                    "delete": {"operationId": f"delete{o.name}", "responses": {"204": {"description": "Deleted"}}},
                }
            doc = {"openapi": "3.1.0", "info": {"title": self.title, "version": "0.0.0"},
                   "paths": paths, "components": {"schemas": schemas}}
            return EmittedFile(path="openapi.json", content=_dump(doc))

        return per_model(whole)(ctx)


# The module:symbol targets: an INSTANCE, or a FUNCTION returning one (called for you with no
# arguments). Not the class itself — a class has a `generate` attribute, so it would be taken
# as the generator and called unbound.
def json_schema() -> JsonSchemaGenerator:
    return JsonSchemaGenerator()


def openapi() -> OpenApiGenerator:
    return OpenApiGenerator(title="Shop")

File: docs/recipes/generators/typescript/json-schema.ts
// EXAMPLE GENERATOR — copy it into your repo (codegen/generators/json-schema.ts) and own it.
// It is a worked example, NOT a supported MetaObjects product surface: no release promises
// its output, and when you need a different shape you edit your copy.
//
// Wire it in metaobjects.config.ts:
//   import { jsonSchemaFile } from "./codegen/generators/json-schema.js";
//   export default defineConfig({ generators: [jsonSchemaFile()] });
//
// emits:  <outDir>/schemas/<package path>/<Name>.schema.json — one JSON Schema (draft
//         2020-12) per concrete object in the model (entities, value objects, projections).
//         `meta verify --codegen` re-runs it and fails when a committed schema is stale.
//
// The three things every generator gets right, all visible below:
//   1. WHICH objects — `ctx.entities` is every object in the model, abstract bases
//      included, so filter with `isAbstract` (or another exported predicate).
//   2. RESOLVING accessors (ADR-0039) — `fields()`, `attr()`, `isRequired`, `maxLength`,
//      `resolvedIsArray()` all see what a field inherits through `extends`. The `own*()`
//      forms and the raw `isArray` flag do not; do not use them here.
//   3. Names from the model — the object's own `name`, its package via `effectivePackage`,
//      and `@objectRef` resolved with `objectRefTarget` (package-aware, never a bare
//      string match).
import {
  effectivePackage,
  enumValues,
  isAbstract,
  objectRefTarget,
  packageToPath,
  perEntity,
  type Generator,
} from "@metaobjectsdev/codegen-ts";
import type { MetaField, MetaObject } from "@metaobjectsdev/metadata";
import {
  DOC_ATTR_DESCRIPTION,
  FIELD_ATTR_LOCAL_TIME,
  FIELD_SUBTYPE_BOOLEAN,
  FIELD_SUBTYPE_CURRENCY,
  FIELD_SUBTYPE_DATE,
  FIELD_SUBTYPE_DECIMAL,
  FIELD_SUBTYPE_DOUBLE,
  FIELD_SUBTYPE_ENUM,
  FIELD_SUBTYPE_FLOAT,
  FIELD_SUBTYPE_INET,
  FIELD_SUBTYPE_INT,
  FIELD_SUBTYPE_LONG,
  FIELD_SUBTYPE_MAP,
  FIELD_SUBTYPE_OBJECT,
  FIELD_SUBTYPE_TIME,
  FIELD_SUBTYPE_TIMESTAMP,
  FIELD_SUBTYPE_URI,
  FIELD_SUBTYPE_UUID,
} from "@metaobjectsdev/metadata/constants";

export type JsonSchema = { [key: string]: unknown };

export interface JsonSchemaFileOpts {
  /** Narrow which objects get a schema (ANDed with "not abstract"). */
  filter?: (entity: MetaObject) => boolean;
  /** Named output target, as on every generator. */
  target?: string;
}

/** Where an object's schema lands, relative to the target's outDir. */
export function schemaPath(obj: MetaObject): string {
  const pkg = effectivePackage(obj);
  const dir = pkg === undefined ? "schemas" : `schemas/${packageToPath(pkg)}`;
  return `${dir}/${obj.name}.schema.json`;
}

/** Relative `$ref` from one object's schema file to another's. */
function relativeRef(from: MetaObject, to: MetaObject): string {
  const fromDir = schemaPath(from).split("/").slice(0, -1);
  const toParts = schemaPath(to).split("/");
  let i = 0;
  while (i < fromDir.length && fromDir[i] === toParts[i]) i++;
  const up = fromDir.slice(i).map(() => "..");
  const rel = [...up, ...toParts.slice(i)].join("/");
  return up.length === 0 ? `./${rel}` : rel;
}

/**
 * The JSON Schema for ONE field's value. `refFor` decides how a `field.object` points at its
 * target — a file-relative `$ref` here, a `#/components/schemas/...` pointer in openapi.ts —
 * which is the one thing the two documents disagree about.
 *
 * The encodings follow the cross-port wire contract (docs/features/api-contract.md, "Type
 * encodings"): decimals travel as strings, currency as integer minor units, temporals as
 * ISO strings.
 */
export function fieldSchema(field: MetaField, refFor: (target: MetaObject) => string): JsonSchema {
  let value: JsonSchema;
  switch (field.subType) {
    case FIELD_SUBTYPE_INT:
    case FIELD_SUBTYPE_LONG:
    case FIELD_SUBTYPE_CURRENCY:
      value = { type: "integer" };
      break;
    case FIELD_SUBTYPE_DOUBLE:
    case FIELD_SUBTYPE_FLOAT:
      value = { type: "number" };
      break;
    case FIELD_SUBTYPE_DECIMAL:
      value = { type: "string", pattern: "^-?\\d+(\\.\\d+)?$" };
      break;
    case FIELD_SUBTYPE_BOOLEAN:
      value = { type: "boolean" };
      break;
    case FIELD_SUBTYPE_DATE:
      value = { type: "string", format: "date" };
      break;
    case FIELD_SUBTYPE_TIME:
      value = { type: "string", format: "time" };
      break;
    case FIELD_SUBTYPE_TIMESTAMP:
      // A @localTime timestamp is a naive wall clock (no `Z`), which `date-time` rejects.
      value = field.attr(FIELD_ATTR_LOCAL_TIME) === true
        ? { type: "string" }
        : { type: "string", format: "date-time" };
      break;
    case FIELD_SUBTYPE_UUID:
      value = { type: "string", format: "uuid" };
      break;
    case FIELD_SUBTYPE_URI:
      value = { type: "string", format: "uri" };
      break;
    case FIELD_SUBTYPE_INET:
      value = { type: "string", anyOf: [{ format: "ipv4" }, { format: "ipv6" }] };
      break;
    case FIELD_SUBTYPE_ENUM:
      value = { type: "string", enum: enumValues(field) ?? [] };
      break;
    case FIELD_SUBTYPE_MAP:
      value = { type: "object" };
      break;
    case FIELD_SUBTYPE_OBJECT: {
      const target = objectRefTarget(field);
      value = target === undefined ? { type: "object" } : { $ref: refFor(target) };
      break;
    }
    default:
      value = { type: "string" };
  }
  if (field.maxLength !== undefined) value.maxLength = field.maxLength;

  // resolvedIsArray(), never the raw `isArray` flag: an array-ness inherited through
  // `extends` lives on the parent field.
  const schema: JsonSchema = field.resolvedIsArray() ? { type: "array", items: value } : value;
  const description = field.attr(DOC_ATTR_DESCRIPTION);
  if (typeof description === "string") schema.description = description;
  // A derived (origin.*) field is computed by the database; clients never send it.
  if (field.isDerived()) schema.readOnly = true;
  return schema;
}

/** The `{ properties, required }` body shared by the file schema and the OpenAPI component. */
export function objectSchema(obj: MetaObject, refFor: (target: MetaObject) => string): JsonSchema {
  const properties: Record<string, JsonSchema> = {};
  const required: string[] = [];
  // fields() RESOLVES: an entity that `extends` BaseEntity gets BaseEntity's fields too.
  for (const field of obj.fields()) {
    properties[field.name] = fieldSchema(field, refFor);
    if (field.isRequired) required.push(field.name);
  }
  const schema: JsonSchema = { title: obj.name, type: "object", properties };
  if (required.length > 0) schema.required = required;
  const description = obj.attr(DOC_ATTR_DESCRIPTION);
  if (typeof description === "string") schema.description = description;
  return schema;
}

export function jsonSchemaFile(opts: JsonSchemaFileOpts = {}): Generator {
  const generator: Generator = {
    name: "json-schema",
    // Abstract bases contribute shape through `extends`; they are not documents of their own.
    filter: (obj) => !isAbstract(obj) && (opts.filter?.(obj) ?? true),
    generate: perEntity((obj) => ({
      path: schemaPath(obj),
      content: `${JSON.stringify(
        {
          $schema: "https://json-schema.org/draft/2020-12/schema",
          $id: schemaPath(obj),
          ...objectSchema(obj, (target) => relativeRef(obj, target)),
        },
        null,
        2,
      )}\n`,
    })),
  };
  if (opts.target !== undefined) generator.target = opts.target;
  return generator;
}
